cupy/core/_routines_math.pyx

```diff
@@ -213,35 +368,117 @@ cdef ndarray scan(ndarray a, ndarray out=None):
+
+    block_size = 512
+    # Since we need to pad each batch we spawn more threads as some
+    # of them will be idle
+    # Calc the total number of blocks
+    padded_bs = 1 << ((batch_size - 1).bit_length())
+    if padded_bs > block_size:
+        blocks_per_batch = (batch_size - 1) // block_size + 1
+        padded_bs = block_size * blocks_per_batch
+    padded_size = a.size // batch_size * padded_bs
+
+    kern_scan = _inclusive_batch_scan_kernel(a.dtype, block_size, op)
+    kern_scan(grid=((padded_size - 1) // (block_size) + 1,),
+              block=(block_size,),
+              args=(a, out, batch_size),
+              shared_mem=a.itemsize * block_size)
+    if batch_size > block_size:
+        blocked_sum = out[:, block_size-1::block_size]
+        _batch_scan_op(blocked_sum, op, dtype, blocked_sum)
+        kern_add = _add_scan_batch_blocked_sum_kernel(
+            out.dtype, op, block_size)
+        kern_add(
+            grid=((out.size - 1) // (block_size) + 1,),
+            block=(block_size,),
+            args=(out, batch_size))
+    return out
+
+
+cdef _axis_to_first(ndarray x, axis):
+    if axis < 0:
+        axis = x.ndim + axis
+    trans = [axis] + [a for a in six.moves.range(x.ndim) if a != axis]
+    pre = list(six.moves.range(1, axis + 1))
+    succ = list(six.moves.range(axis + 1, x.ndim))
+    revert = pre + [0] + succ
+    return trans, revert
+
+
+cdef _proc_as_batch(ndarray x, axis, dtype, op):
+    if x.shape[axis] == 0:
+        return cupy.empty_like(x)
+    trans, revert = _axis_to_first(x, axis)
+    t = x.transpose(trans)
+    s = t.shape
+    r = t.reshape(x.shape[axis], -1)
+    # This is to use the current implemented fast scan
+    # TODO(ecastill) merge with above transformations
+    r = r.transpose((1, 0))
+    r = _batch_scan_op(r, op, dtype, r)
+    r = r.transpose((1, 0))
+    return r.reshape(s).transpose(revert)
+
+
+cdef _cum_core(ndarray a, axis, op, dtype, ndarray out):
+    if out is None:
+        if dtype is None:
+            kind = a.dtype.kind
+            if kind == 'b':
+                dtype = numpy.dtype('l')
+            elif kind == 'i' and a.dtype.itemsize < numpy.dtype('l').itemsize:
+                dtype = numpy.dtype('l')
+            elif kind == 'u' and a.dtype.itemsize < numpy.dtype('L').itemsize:
+                dtype = numpy.dtype('L')
+            else:
+                dtype = a.dtype
+        out = a.astype(dtype=dtype)
+    else:
+        out[...] = a
+    if axis is None:
+        out = out.ravel()
+        return _scan_op(out, op, dtype, out)
+    elif not (-a.ndim <= axis < a.ndim):
+        raise cupy.core._AxisError('axis(={}) out of bounds'.format(axis))
+    else:
+        return _proc_as_batch(out, axis, dtype, op)
+
+
 # Only for test
 def _scan_for_test(a, out=None):
-    return scan(a, out)
+    return scan(a, scan_op.SCAN_SUM, dtype=None, out=out)
+
+
+def scan_core(a, axis, op, dtype=None, out=None):
+    return _cum_core(a, axis, op, dtype, out)
```

Comment:
Why did you split the function into `scan_core` and `_cum_core`?

Reply:
Because of my weak cython skills 😂